Recommend noc_async_write_flushed() on examples

jaebaek opened the pull request:
As an alternative of `noc_async_write_barrier()`, this commit updates examples to recommend `noc_async_write_flushed()` that can be faster than `noc_async_write_barrier()` because it waits until the write request is sent.

### Ticket
No Github issue, but I met a case that `noc_async_write_barrier()` causes an infinite loop, while `noc_async_write_flushed()` finishes the kernel correctly. You can reproduce it by updating [this line](https://github.com/jaebaek/tenstorrent-tiny-examples/blob/524dee905e4b80a0f5b11bd386d899f047a16315/src/kernels/multicast_advanced.cpp#L98) to `noc_async_write_barrier()`. I explained this issue on [Discord channel](https://discord.com/channels/863154240319258674/1177107133167321158/1272042576664596541).

### Problem description
I met a case that `noc_async_write_barrier()` causes an infinite loop, while `noc_async_write_flushed()` finishes the kernel correctly. You can reproduce it by updating [this line](https://github.com/jaebaek/tenstorrent-tiny-examples/blob/524dee905e4b80a0f5b11bd386d899f047a16315/src/kernels/multicast_advanced.cpp#L98) to `noc_async_write_barrier()`.

### What's changed
I added comments on programming examples for tt-metal to recommend `noc_async_write_flushed()`.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

jaebaek committed:
Recommend noc_async_write_flushed() on examples

As an alternative of `noc_async_write_barrier()`, this commit updates
examples to recommend `noc_async_write_flushed()` that can be faster
than `noc_async_write_barrier()` because it waits until the write
request is sent.

jaebaek commented:
I am curious .. when I can possibly have an infinite loop because of `noc_async_write_barrier()`.
[My example](https://github.com/jaebaek/tenstorrent-tiny-examples/blob/524dee905e4b80a0f5b11bd386d899f047a16315/src/kernels/multicast_advanced.cpp#L98) has that issue. `noc_async_writes_flushed()` fixed the issue, so I used `noc_async_writes_flushed()` instead.

tt-aho commented:
Hey @jaebaek,

I think this note is a good addition to the examples, however, if you use `noc_async_writes_flushed()`, you should still have a final `noc_async_write_barrier()` at the end of your kernel so that all writes have actually completed when the kernel terminates, as our runtime expects kernels to complete all transactions before they terminate. We can add this as part of your comment about using flush. An example of why this might be a problem is because if we don't have the final barrier then your kernel finishing doesn't mean your output data has actually been committed, and if we try to read back the output tensor after the operation executed it's possible the writes have not completed and we read back garbage data.

As for the issue you're facing where you seem to hang using `noc_async_write_barrier()`, it's possible that you are issuing an illegal NOC transaction, for example targeting an incorrect core, so the write never finishes. We have a tool called [watcher](https://docs.tenstorrent.com/tt-metalium/latest/tools/watcher.html#watcher) that can help catch issues like this. You can try running your test with `TT_METAL_WATCHER=1` set you see if it catches any issues.

jaebaek commented:
Thank you for clarifying it :)
I actually realized my example code multicast-sent a tile to the sender itself. I guess that's the root cause of the infinite loop but definitely I want to try the watcher tool.

I updated this PR. I added the comment only on the examples with multiple `noc_async_write_barrier()` calls. I also explained we have to use `noc_async_write_barrier()` at least once at the end of the data move kernel.

jaebaek commented:
By the way, I finally fixed the bug of the infinite loop on `noc_async_write_barrier()` from [my example](https://github.com/jaebaek/tenstorrent-tiny-examples/tree/main/src/5_multicast_advanced). Thanks a lot!

tt-rkim commented:
Closing in favour of https://github.com/tenstorrent/tt-metal/pull/11448

jaebaek commented:
Thanks!
